bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/internal/RuleRegistryImpl.java

```diff
@@ -461,87 +489,343 @@ public Boolean isEnabled(String ruleUID) {
+        if (template == null) {
+            updateRuleTemplateMaping(templateUID, uid, false);
+            RuleStatusInfo statusInfo = disabledRulesStorage != null && disabledRulesStorage.get(uid) == null
+                    ? new RuleStatusInfo(RuleStatus.UNINITIALIZED, RuleStatusDetail.TEMPLATE_MISSING_ERROR)
+                    : new RuleStatusInfo(RuleStatus.DISABLED, RuleStatusDetail.TEMPLATE_MISSING_ERROR);
+            addToRuleEngine(rule, statusInfo);
+            logger.debug("Rule template {} does not exist.", templateUID);
+            return rule;
+        } else {
+            Rule resolvedRule = new Rule(uid, RuleUtils.getTriggersCopy(template.getTriggers()),
+                    RuleUtils.getConditionsCopy(template.getConditions()),
+                    RuleUtils.getActionsCopy(template.getActions()), template.getConfigurationDescriptions(),
+                    rule.getConfiguration(), null, rule.getVisibility());
+            resolvedRule.setName(rule.getName());
+            resolvedRule.setTags(rule.getTags());
+            resolvedRule.setDescription(rule.getDescription());
+            validateModuleIDs(resolvedRule, resolvedRule.getModules(Module.class));
+            resolveConfigurations(resolvedRule);
+            updateRuleTemplateMaping(templateUID, uid, true);
+            return resolvedRule;
+        }
+    }
 
-                return resolvedRule;
+    /**
+     * Updates the content of the {@link Map} that maps the template to rules, using it to complete their definitions.
+     *
+     * @param templateUID the {@link RuleTemplate}'s UID specifying the template.
+     * @param ruleUID the {@link Rule}'s UID specifying a rule created by the specified template.
+     * @param resolved specifies if the {@link Map} should be updated by adding or removing the specified rule
+     *            accordingly if the rule is resolved or not.
+     */
+    private void updateRuleTemplateMaping(String templateUID, String ruleUID, boolean resolved) {
+        synchronized (this) {
+            Set<String> ruleUIDs = mapTemplateToRules.get(templateUID);
+            if (ruleUIDs == null) {
+                ruleUIDs = new HashSet<String>();
+                mapTemplateToRules.put(templateUID, ruleUIDs);
+            }
+            if (resolved) {
+                ruleUIDs.remove(ruleUID);
+            } else {
+                ruleUIDs.add(ruleUID);
             }
         }
-        return rule;
     }
 
     @Override
     protected void addProvider(Provider<Rule> provider) {
         super.addProvider(provider);
         Collection<Rule> rules = new LinkedList<Rule>(elementMap.get(provider));
         for (Rule rule : rules) {
-            updateRuleByTemplate(provider, rule);
+            try {
+                Rule resolvedRule = resolveRuleByTemplate(rule);
+                if (rule != resolvedRule) {
+                    update(resolvedRule);
+                }
+            } catch (Exception e) {
```

Comment:
We should not catch generic Exceptions. Can you make this more specific?

Reply:
OK